core/src/term/mod.rs

```diff
@@ -261,6 +261,13 @@ pub enum Term {
     ///
     /// This is a temporary solution, and will be removed in the future.
     Closure(CacheIndex),
+
+    #[serde(skip)]
+    /// An opaque value that cannot be constructed within Nickel code.
+    ///
+    /// This can be used by programs that embed Nickel, as they can inject these opaque
+    /// values into the AST.
+    Opaque(u64),
```

Comment:
I don't know if we spell `u64` somewhere else, but I wonder if we should use a type alias instead to make it easy to switch to something else (ie define `Opaque(OpaquePayload)` with `type OpaquePayload = u64` or something).

Reply:
It might even be fairly straightforward to make the interpreter generic over these if desired; I don't actually need that for my usecases so it might not be worth it, but it might allow generalizing to other systems nicely.

Reply:
I think that would be a bit painful, since it would involve carrying around an extra type parameter everywhere that `Term` is used (which is basically everywhere). An easier version would be to have a `Box<dyn Any>` to allow downcasting.

Reply:
I agree with Joe, on paper it's just adding a generic, but API-wise, this is painful, because you often have to propagate this type parameter everywhere (every site that uses a term, which is pretty much everywhere). For less fundamental datastructure we could a parameter and make an alias like `pub type Foo = ParametrizedFoo<usize>`, but for terms, I'm afraid this will leak in `typechecking`, `transform`, etc. which might not be worth it. The `Box<dyn Any>` doesn't look very nice, but yeah, if it's really needed, this might be a possibility.